src/Core/Transmission.Abstractions/FileStorage.cs

```diff
@@ -51,36 +60,29 @@ protected FileStorage(string storagePath)
         /// </summary>
         protected abstract string EncodeFileName(TData data);
 
-        /// <inheritdoc/>
-        public bool HasData => _files.Any();
-
         /// <inheritdoc/>
         public async Task<IReadOnlyCollection<TData>> DequeueAsync(
             int count,
             CancellationToken cancellationToken)
         {
-            if (!HasData)
-            {
-                return EmptyBatch;
-            }
-
             var batch = new List<TData>();
 
-            foreach(FileInfo file in _files.Take(count))
+            for (var i = 0; i < count; i++)
             {
-                var fileName = Path.GetFileNameWithoutExtension(file.FullName);
+                var file = await _filesReader.ReadAsync(cancellationToken);
+                var fileName = Path.GetFileNameWithoutExtension(file);
 
                 using (await FilesLock.ReadLockAsync(fileName, cancellationToken))
```

Comment:
Do we need a lock on the file? I mean after it is enqueued can there be changes to it?